rfcs/20230110-kernel-reuse/README.md

```diff
@@ -0,0 +1,300 @@
+
+Method already exists within oneDNN so there is already precedent.
+
+Feature has been requested by frameworks
+
+Implementable for all kernel generation methods
+
+#### Cons:
+Makes API more complicated
+
+Requires user source code changes to benefit
+
+Increases testing load as we need to test combinations of runtime parameters.
+
+Does not improve benchdnn testing performance
+
+Unless care is taken, this will lead to more implementations, increasing
+implementation and maintenance burden
+
+Opaque user experience. The user needs to choose between using runtime
+parameters to reuse a kernel but with the trade off that they may get worse
+performance. Since performance is implementation dependent, users are in a bad
+position to when they can make this choice with limited performance degradation.
+
+More work as API support should be implemented for both CPU and GPU.
+
+#### Implementation:
+The implementation will be highly dependent on primitive implementations. There
+are two main scenarios here. First scenario, we modify our implementations to
+generate a finite number of kernels. We then use runtime parameters to
+dynamically switch top the best implementation. In this case, since we are
+reusing the same finite number of kernels, maintainability is not impacted. The
+other scenario requires a separate implementation which increases testing
+requirements and maintenance burden.
+
+### Reuse Option 3: Fixed Size Runtime Function Generator Registries
+This option creates a fixed size registry for storing reusable kernel
+generators. These generators are expected to be performant. For OpenCL C
+kernels, these are expected to just be program binaries.
+
+#### Pros:
+Requires no user intervention.
+
+No API changes.
+
+Relatively simple implementation
+
+Requirements on Function Generators is similar to Distribution Option 1.
+
+#### Cons:
+Requires a finite number of generators, no guarantee generators align with
+customer workloads.
+
+Users do not control the registry size
+
+Requires the generator to be GPU agnostic (i.e. should work for 2 or more GPUs
+in a system if they have the same architecture) to limit the registry size.
+
+Implementation needs to limit maximum memory usage to a reasonable level.
+
+#### Implementation
+This feature will be completely managed by oneDNN developers. As such, there are
+no API changes. The only guarantee is that the registry is generated at runtime
+and will not use much memory until users creates primitives. It is expected the
+implementation will be a hashmap between a kernel configuration and a kernel
+generator. Registries will have a compile time defined maximum size to avoid
+excessive memory usage. As a registries size is dependent on implementation,
+separate registries will exist per implementation. As the requirements on this
+implementations are similar to Distribution Option 1,this can be used to enable
+build time compilation.
+
+### Reuse Option 4: Global Kernel Generator Cache
```

Comment:
I agree that this option looks like the best tradeoff. While the section has a sufficient high-level description of the option it lacks certain level of details about the design of the feature.
I think at least the following aspect should be covered:
* Definition of the key
* Definition of the value
* How/where load/store operations are performed
* Any data about primitive (`primitive_t`) creation cost without kernel creation. We need to make sure that the overhead is acceptable

Please see the following RFCs for reference:
* [Persistent cache](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20210624-persistent-cache)
* [Optimization of primitive descriptor creation](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20210129-optimize-pd-creation)
* [Library size reduction](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20210510-binary-size-reduction)

A final note, when creating an RFC we usually try to provide details on design for all suggested options. I don't think it makes sense to do it now but I just wanted to mention it for future RFCs.